File: javascript/observer/NewsLetter/subject/Newsletter.js
import {Subject} from "./Subject.js";
import {Observer} from "../observer/Observer.js";

export class Newsletter extends Subject{

    #name
    #observers

    constructor() {
        super();
        this.#observers = []
    }

    registerObserver(newObserver) {
        super.registerObserver(newObserver);
        this.#observers.push(newObserver)
    }

    removeObserver(removeObserver) {
        super.removeObserver(removeObserver);
        let index = this.#observers.indexOf(removeObserver)
        this.#observers.splice(index, 1)
    }

    notifyObservers() {
        super.notifyObservers();
        let observer = new Observer()
        for (observer of this.#observers) {
            observer.update(this.#name)
        }
    }

    setName(name) {
        this.#name = name
        this.notifyObservers()
    }

}

File: ruby/observer/NewsLetter/subject/newsletter.rb
require './subject/subject'
require './observer/observer'

class Newsletter < Subject

  def initialize
    super
    @observers = []
    @name = ''
    @observer = Observer.new
  end

  def register_observer(new_observer)
    super
    @observers.push(new_observer)
  end

  def remove_observer(remove_observer)
    super
    @observers.delete(remove_observer)
  end

  def notify_observers
    super
    for @observer in @observers
      @observer.update(@name)
    end
  end

  def set_name(name)
    @name = name
    notify_observers
  end

end